homeassistant/components/sensor/thermoworks_smoke.py

```diff
@@ -0,0 +1,161 @@
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_EXCLUDE, default=[]):
+        vol.All(cv.ensure_list),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the thermoworks sensor."""
+    import thermoworks_smoke
+    from requests.exceptions import HTTPError
+
+    email = config.get(CONF_EMAIL)
+    password = config.get(CONF_PASSWORD)
+    monitored_variables = config.get(CONF_MONITORED_VARIABLES)
+    excluded = config.get(CONF_EXCLUDE)
+
+    try:
+        mgr = thermoworks_smoke.initialize_app(email, password, True, excluded)
+
+        # list of sensor devices
+        dev = []
+
+        # get list of registered devices
+        for serial in mgr.serials():
+            for variable in monitored_variables:
+                dev.append(ThermoworksSmokeSensor(variable, serial, mgr))
+
+        add_devices(dev, True)
+    except HTTPError as error:
+        msg = "{}".format(error.strerror)
+        if 'EMAIL_NOT_FOUND' in msg or \
+                'INVALID_PASSWORD' in msg:
+            _LOGGER.error("Invalid email and password combination")
+        else:
+            _LOGGER.error(msg)
+
+
+class ThermoworksSmokeSensor(Entity):
+    """Implementation of a thermoworks smoke sensor."""
+
+    def __init__(self, sensor_type, serial, mgr):
+        """Initialize the sensor."""
+        self._name = "{name} {sensor}".format(name=mgr.name(serial),
+                                              sensor=SENSOR_TYPES[sensor_type])
+        self.type = sensor_type
+        self._state = None
+        self._attributes = {}
+        self._unit_of_measurement = TEMP_FAHRENHEIT
+        self.serial = serial
+        self.mgr = mgr
+        self.update_unit()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this sensor."""
+        return self._unit_of_measurement
+
+    def update_unit(self):
+        """Set the units from the data."""
+        if PROBE_2 in self.type:
+            self._unit_of_measurement = self.mgr.units(self.serial, PROBE_2)
+        else:
+            self._unit_of_measurement = self.mgr.units(self.serial, PROBE_1)
+
+    def update(self):
+        """Get the monitored data from firebase."""
+        try:
+            values = self.mgr.data(self.serial)
+
+            # set state from data based on type of sensor
+            self._state = values.get(self.type)
+
+            # set units
+            self.update_unit()
+
+            # set basic attributes for all sensors
+            self._attributes = {
+                'serial': self.serial,
```

Comment:
We don't store serial number, firmware version and similar data in state attributes. Soon we can use device info property to report this. State attribute should hold data that changes as the state of the device changes.

Reply:
Where can I put it so the user can get at it? The serial number is an important unique identifier for their device. I can find plenty of examples of other components that DO store that data as attributes (or at least they show up in the Attributes column in the state viewer).

Reply:
Where in the process of implementing `device_info` entity property, and a device registry. It's still in flux though, so we can't use it here right now. It might also be restricted to components that have implemented config entries.

It's probably a good idea to use the serial for the `unique_id` property. That way users will be able to easily change name and entity_id of the entities.

Reply:
I'm unfamiliar with that property. can you point me to the documentation of special properties?

Reply:
https://developers.home-assistant.io/docs/en/entity_index.html#generic-properties

https://developers.home-assistant.io/docs/en/entity_registry_index.html

Note that defining a unique_id won't mean that users have access to the serial number in state attributes.

Reply:
Is it ok for the unique_id to have colons in it? It looks like a mac address. Presumably I would append the sensor name so each one was like 00:00:00:00:00:00-probe1 ...-probe2 and so on.

Reply:
It can be any string, but should be unique per entity and constant. Yes, adding sensor type to the serial is good way of making a unique id if the platform generates multiple sensors per device.

Reply:
Static attributes not related to state should be removed from state attributes.

Reply:
Is device_info ready to use. this information needs to be somewhere.

Reply:
Yes, but it requires using config entries. Please add that in a separate PR.

https://developers.home-assistant.io/docs/en/device_registry_index.html

Reply:
added device info